src/mapping/parser/mod.rs

```diff
@@ -0,0 +1,293 @@
+                for option in segment.into_inner() {
+                    match option.as_rule() {
+                        Rule::path_segment => options.push(option.as_str().to_string()),
+                        _ => unreachable!(),
+                    }
+                }
+                segments.push(options);
+            }
+            _ => unreachable!(),
+        }
+    }
+    Ok(segments)
+}
+
+fn query_from_pair(pair: Pair<Rule>) -> Result<Box<dyn query::Function>> {
+    Ok(match pair.as_rule() {
+        Rule::string => Box::new(Literal::from(Value::from(
+            pair.into_inner()
+                .next()
+                .unwrap()
+                .as_str()
+                // TODO: Include unicode escape sequences, surely there must be
+                // a standard lib opposite of https://doc.rust-lang.org/std/primitive.str.html#method.escape_default
+                // but I can't find it anywhere.
+                .replace("\\\"", "\"")
+                .replace("\\n", "\n")
+                .replace("\\t", "\t")
+                .replace("\\\\", "\\"),
+        ))),
+        Rule::null => Box::new(Literal::from(Value::Null)),
+        Rule::number => Box::new(Literal::from(Value::from(
+            pair.as_str().parse::<f64>().unwrap(),
+        ))),
+        Rule::boolean => {
+            let v = if pair.as_str() == "true" { true } else { false };
+            Box::new(Literal::from(Value::from(v)))
+        }
+        Rule::dot_path => Box::new(QueryPath::from(path_segments_from_pair(pair)?)),
+        _ => unreachable!(),
+    })
+}
+
+fn mapping_from_pairs(pairs: Pairs<Rule>) -> Result<Mapping> {
+    let mut assignments = Vec::<Box<dyn Function>>::new();
+    for pair in pairs {
+        match pair.as_rule() {
+            Rule::assignment => {
+                let mut inner_rules = pair.into_inner();
+                let path = path_from_pair(inner_rules.next().unwrap())?;
+                let query = query_from_pair(inner_rules.next().unwrap())?;
+                assignments.push(Box::new(Assignment::new(path, query)));
+            }
+            Rule::deletion => {
+                let mut inner_rules = pair.into_inner();
+                let path = path_from_pair(inner_rules.next().unwrap())?;
+                assignments.push(Box::new(Deletion::new(path)));
+            }
+            _ => (),
+        }
+    }
+    Ok(Mapping::new(assignments))
+}
+
+pub fn parse(input: &str) -> Result<Mapping> {
+    match MappingParser::parse(Rule::mapping, input) {
+        Ok(a) => mapping_from_pairs(a),
+        Err(err) => Err(format!("mapping parse error\n{}", err)),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn check_parser_errors() {
+        let cases = vec![
+            (
+                ".foo = {\"bar\"}",
+                r###"mapping parse error
+ --> 1:8
+  |
+1 | .foo = {"bar"}
+  |        ^---
+  |
+  = expected dot_path, boolean, null, string, or number"###,
```

Comment:
I would expect changes in the pest crate to make these parser errors fragile. Maybe test for key strings (perhaps the location and error message) instead of an exact string match.